Williams homework 1 (#5)
timsexperiments/intro-to-relational-databases: submissions/homework_1/cc1759/hw.md

### **Amazon**

Amazon probally databases for your lists, for your recomendations, for your past orders, for all there products, a list of sellers, and your cart.

For products they should have a rational data base, where the row is the product id, and the columnns are data like product name, product tags, item price, amnt in stock, the seller, its ratings, its description, product information (in JSON which i assumeing you can attach to a table in sql), and the reviews. Since I have nothing for listed for the table, have each table be a catagory so search is easier. Theres also should be a thing so you can change type of the object (like changing its color etc..), which woudl change a few of the things on the table like the products price and maybe the product details I'm not sure, but I havn't thought about how to do that yet. It would look somethign like:

**Table Name**
Catagory

**Rows *(primary key)***
- Product_ID: VARCHAR or INT AUTO_INCREMENT 

**Columns**
- Name: TEXT
- Author: VARCHAR
- Seller_ID: VARCHAR

> - Tags: I would say set, but I think that would mean you have to list every possible tag for each product, and would have a max of 64 tags so im not sure.

Author: Yeah, literally when i got to lesson 2 a lot of my questions i answered and looked back on how i would do it lol.

For your list database I would go with a *Key-Value Database*, with each key being the name of the list and the value being a list of the product values ID's, which are then looked up and grabbed from the products table. Though one thing I will say, is that they are ordered by last placed in list and IDK if you can do that with Key_Value pairs but idk. Otherwise it could be a relational database with table being the list, row being the an auto incramented index, and the collums having the products ID. Then its also expandable to any other info you'd want to add like date added to the list or whatever.

For recomendations, I think they would go with a *Vector Database* to search for similar products to what you've bought in the past, recently searched for, or have saved in your lists.

> Your cart would either be more simaler to the past orders database or the list database I'm not sure. I would think the lists database because I don't think it would need all the extra information like the past orders table, so maybe litterally just make it the same as the lists

Author: Cool.

> Your cart would either be more simaler to the past orders database or the list database I'm not sure. I would think the lists database because I don't think it would need all the extra information like the past orders table, so maybe litterally just make it the same as the lists

Author: nerd

> Your cart would either be more simaler to the past orders database or the list database I'm not sure. I would think the lists database because I don't think it would need all the extra information like the past orders table, so maybe litterally just make it the same as the lists

Author: jk

they probally also have a list for items ordered by everyone so they can ship the right things out to the right location.

### **Discord**

Discord probally has databases direct messages, for your friends, for your joined servers, server messages, and for users and there settings.

For users it would probally have a *Relational Database*, with rows being your user ID (VARCHAR) and columns being things such as, disp_name, status, Tiene_Nitro, date joined, and all the settings (or they would have it)

> For severs joined, I would have to choose a *Graph Database*. Have each user and each server be a node, and have a user be connected to the servers they've joined by node. This should provide for easy lookup for mutual server with other users (though im not sure how the lookup would be any faster than a list?)

Reviewer: Typos:

```suggestion
For severs joined, I would have to choose a *Graph Database*. Have each user and each server be a node, and have a user be connected to the servers they've joined by node. This should provide for easy lookup for mutual server with other users (though I'm not sure how the lookup would be any faster than a list?)
```

> For friend, I would have it be a *Graph Database* very simaler to the joined servers database, but with all the nodes being users, and the edges connecting those who are friends.

Reviewer: typos:

```suggestion
For friends, I would have it be a *Graph Database* very similar to the joined servers database, but with all the nodes being users, and the edges connecting those who are friends.
```

> For private messaging and server messages, it would probally be a JSON each JSON itself hold data for a server. In it are messages as objects with different properties (author, time posted, message contents) nested within which channel its in. Since the server/DM messaging and channels are probally in JSON format, I would us a *Document Database* 

Reviewer: Typos:

```suggestion
For private messaging and server messages, it would probably be a JSON each JSON itself hold data for a server. In it are messages as objects with different properties (author, time posted, message contents) nested within which channel its in. Since the server/DM messaging and channels are probably in JSON format, I would us a *Document Database*
```

> ## **Part 2: Define a Rational Database**

Reviewer: "Rational" -> "Relational"

```suggestion
## **Part 2: Define a Rational Database**
```
